### ISSUE TITLE
help on changing variables in train loop

### ISSUE THREAD

**Issue by level14taken** [pytorch/xla#1865]
```
def train_loop_fn(loader):
    model.train()
    num = 20*len(trn_loader)-1
    mult = (final_value / init_value) ** (1/num)
    lr = init_value
    for x, (image, mask) in enumerate(loader): 
      y_pred = model(image)
      loss = loss_fn(y_pred, mask)
      optimizer.zero_grad()
      loss.backward()
      xm.optimizer_step(optimizer)
      if xm.is_master_ordinal():
        global uti
        uti['batch_num'] += 1
        uti['avg_loss'] = beta * uti['avg_loss']  + (1-beta) * loss.item()
        smoothed_loss = uti['avg_loss'] / (1 - beta**uti['batch_num'])
        if uti['batch_num'] > 1 and smoothed_loss > 4 * uti['best_loss']:
            return None
        if smoothed_loss < uti['best_loss'] or uti['batch_num']==1:
            uti['best_loss'] = smoothed_loss
        uti['losses'].append(smoothed_loss)
        uti['log_lrs'].append(math.log10(lr))
        lr *= mult
        optimizer.param_groups[0]['lr'] = lr
    return loss.item(),IOU
```
Hi can you help me to use the above code,the problem is that i can't  update the global dict 'uti', as multiprocessing doesen't allow updating global variables.can i make any change in the code that can update the uti dictionary,the original code  was from [here](https://sgugger.github.io/how-do-you-find-a-good-learning-rate.html) I was trying to make it work on tpu

**Comment by dlibenzi** [comment]
You cannot have global variables with multi-processing.
You either run single core, or you compute per core and average.
You could use `xm.rendezvous()` to exchange the data but IMHO a per core estimation is enough:

https://github.com/pytorch/xla/blob/4a4d19dca92d1d31824146779f800362501bd796/torch_xla/core/xla_model.py#L535

**Comment by level14taken** [comment]
> you compute per core and average.
Can you point to a resource or show how this is done.

**Comment by dlibenzi** [comment]
Just run one core by setting `nprocs=1` in your equivalent of our example here:

https://github.com/pytorch/xla/blob/4a4d19dca92d1d31824146779f800362501bd796/test/test_train_mp_mnist.py#L185

And declare your variables not as globals, but within the train function.
Like this:

https://github.com/pytorch/xla/blob/4a4d19dca92d1d31824146779f800362501bd796/test/test_train_mp_mnist.py#L119

**Comment by level14taken** [comment]
let's say i want to store all the losses of the tpu master core  throughout the training and store them in an array, how should i proceed,I intend to use the array after training.

**Comment by dlibenzi** [comment]
Losses are usually very close to each other from the different cores, so getting them from all the core might not be needed.
Also, calling loss.item() at every step will make training pretty slow.
We usually suggest to print/report/save the loss once every N steps:

https://github.com/pytorch/xla/blob/44429d0d0f16979d831c5a21054abf9bfd86653b/test/test_train_mp_mnist.py#L129

If you really want to get the values from each core, you can use something like:
```Python
import pickle
def train_loop(...):
  losses = []
  for ...:
    losses.append(loss.item())
  all_losses = [pickle.loads(x) for x in xm.rendezvous('all_losses', pickle.dumps(losses))]
```
